### PR TITLE
Use dispatch_queue to sync access to db

### PR DESCRIPTION
**Pull request by joanromano**
TODO before merging
- [x] Add unit tests
- [x] Use Box, see #17

### CODE REVIEW

**Comment by MP0w** [comment]
Yep!

**Comment by zzarcon** [comment]
@joanromano really cool man!

File: Source/KakapoDB.swift [hunk 1]
```diff
@@ -17,47 +17,95 @@ enum KakapoDBError: ErrorType {
...
+    func create<T: KStorable>(_: T.Type, number: Int) -> [KStorable] {
+        var result: [KStorable] = []
+        
+        dispatch_barrier_sync(queue) { [weak self] in
+            guard let weakSelf = self else { return }
+            
+            let arrayBox = weakSelf.lookup(T)
+            result = (0..<number).map { _ in T(id: weakSelf.uuid()) }
+            arrayBox.value.appendContentsOf(result)
+            weakSelf.store[String(T)] = arrayBox
```
Comment (MP0w) [inline Source/KakapoDB.swift]: No need to get it and set it back to the dictionary (that also mutate it),  right?

Comment (joanromano) [inline Source/KakapoDB.swift]: Yes since lookup() would either give the entry or a fresh new ArrayBox

Comment (joanromano) [inline Source/KakapoDB.swift]: Should lookup() also store the new ArrayBox before returning it?

Comment (MP0w) [inline Source/KakapoDB.swift]: And when you create the fresh you should also store it

Comment (MP0w) [inline Source/KakapoDB.swift]: +            let arrayBox = weakSelf.lookup(T)
+            result = (0..<number).map { _ in T(id: weakSelf.uuid()) }
+            arrayBox.value.appendContentsOf(result)
+            weakSelf.store[String(T)] = arrayBox

`weakSelf.lookup(T).value.append....`

File: KakapoExample/KakapoTests/KakapoDBTests.swift [hunk 2]
```diff
@@ -55,6 +55,64 @@ class KakapoDBTests: QuickSpec {
...
+                
+                dispatch_apply(5000, queue, { i in
+                    sut.create(CommentFactory.self, number: 1)
+                })
+                
+                let userObjects = sut.findAll(UserFactory.self)
+                let user = sut.find(UserFactory.self, id: 1)
+                
+                let commentObjects = sut.findAll(CommentFactory.self)
+                let comment = sut.find(CommentFactory.self, id: 1002)
```
Comment (MP0w) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: how can you be sure there will be this comment? I don't think is a good idea to mix this 2 things. If comment is fast then a User could take the 1002 id

Comment (MP0w) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: I would just create 5000 comments, they are concurrent anyway

File: KakapoExample/KakapoTests/KakapoDBTests.swift [hunk 3]
```diff
@@ -78,41 +136,47 @@ class KakapoDBTests: QuickSpec {
...
             it("should fail when inserting invalid id", closure: {
-                do {
-                    try sut.insert(UserFactory(firstName: "Joan", lastName: "Romano", age:25, id: 100))
-                    try sut.insert(UserFactory(firstName: "Joan", lastName: "Romano", age:25, id: 100))
-                } catch KakapoDBError.InvalidId {
-                    expect(true).to(beTrue())
-                } catch {
-                    expect(false).to(beTrue())
-                }
+                sut.insert({ (id) -> UserFactory in
```
Comment (MP0w) [inline KakapoExample/KakapoTests/KakapoDBTests.swift:156]: this 2 lines can be removed are not part of this test

File: KakapoExample/KakapoTests/KakapoDBTests.swift (hunk 2 above)
Comment (joanromano) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: Well, for the test it should be enough. Recall that `dispatch_apply` waits until all executions are finished to continue the execution.

Comment (joanromano) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: You can test it for yourself even for smaller numbers

```
             let queue = dispatch_queue_create("queue", DISPATCH_QUEUE_CONCURRENT)
                dispatch_apply(500, queue, { i in
                    sut.create(UserFactory.self, number: 1)
                })

                dispatch_apply(10, queue, { i in
                    sut.create(CommentFactory.self, number: 1)
                })
                let commentObjects = sut.findAll(CommentFactory.self)
                let comment = sut.find(CommentFactory.self, id: 501)
                expect(comment).toNot(beNil())
                expect(comment?.text).toNot(beNil())
                expect(comment?.id) == 501
                expect(commentObjects.count) == 10
```

Should always pass

Comment (MP0w) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: oh :+1: didn't understood that `dispatch_apply` was synchronous, fUck!

Comment (MP0w) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: Then why don't you use 1000 instead if 1002?

Comment (joanromano) [inline KakapoExample/KakapoTests/KakapoDBTests.swift]: I am just retrieving a comment so it doesn't matter from the range [1000...6000]

File: KakapoExample/KakapoTests/KakapoDBTests.swift (hunk 3 above)
Comment (joanromano) [inline KakapoExample/KakapoTests/KakapoDBTests.swift:156]: I think the point is a normal insert works as expected and then a bad insertion is done -> should throw.

By the way I am already subscribed to https://github.com/Quick/Nimble/pull/248

**Pull request merged**